### ISSUE TITLE
[Question] All elements of "velocity" and "load" optional

### ISSUE THREAD

**Issue by leierseder-jh** [VDA5050/VDA5050#96]
As of now, all elements of the "velocity" object are marked as optional:
![image](https://user-images.githubusercontent.com/47557696/155203671-19299979-0b83-4d81-af68-dfdeb594736e.png)

Of course, the whole element "velocity" inside the topics "state" and "visu" can be optional, but shouldn't at least vx and vy be mandatory?

The same applies for the internal structure of "load", where all elements are optional:
![image](https://user-images.githubusercontent.com/47557696/155204078-c1042929-d82f-4216-a9cc-d5dcb9cc33b5.png)

IMHO at least loadId should be mandatory.

**Comment by AntonDueck** [comment]
for vehicles do not know the load Id because thre is no sensor for this und the FMS do not send the LoadId inside the handling action what will happen there? 

vy is only valid vor omnidirectional vehicles, not diffDrive the most common kinematik type. 

is the X and Y the global Koordinate system the Vehicle Center Point ?

if you like to have the global x and y and have an only line Guided agv the  you can not compute the x and y speeds in global. because the agv do not know it position.


The Purpuse of the VDA5050 is  to support the most AGV Types, and the agv with less computional power are the most limiting one.  This limited agv type deside the mondatory parts mostly.

**Comment by leierseder-jh** [comment]
If it is assumed that for the load only values that can be sensed by the vehicle itself are present, I agree that there is at least one scenario for one application where each field cannot be determined.

For the y component of the speed: also true, depending on the definition, kinematic type etc. it might not be possible to determine it.

But at least the x component (be it for a wheel or the center point) has to be present, because if not, *vx*, *vy* and *omega* are not present and thus the whole *velocity* entry can be dropped. That's what I meant: the entry is optional for good reason, but it does not make sense that all 3 elements are missing at the same time then.

**Comment by MaximilianPenninger** [comment]
I don't see the problem. The object velocity in the state-topic is optional too, so for each agv-type the mix of available values is individual. The definition, what is relevant for the agv can be found in the factsheet.
I agree, it's very uncommon that an agv has no x-speed but a y-speed.

**Comment by philipkirmse** [comment]
To bring this discussion back to top. For Load State of the AGV there are a lot of fields which are uncommon but maybe in som cases necessary.

The most common case is not 100% defined. The case would be to know boolean if the AGV is laoded (true/false) or by actual laod capacit eg. (1 of 5)

As I interpret it right this information can only be giffen today by AGV to MC when a list of load positions is send in field loadPosition. If not loaded it stays empty. 

One case why it is necessary, to know actual state if AGV is loaded (which AGV can define itself either by sensors or handling device position). is that an AGV in automatic can be loaded without an Active order. Can happen manualy or orders ar manually deleted. -> So without a actual state you can only handle it by organizational user interaction to take responsibility to worker. If he makes a mistake an new order may be send to AGV which is in real still loaded.

Any suggestions?

**Comment by JWCS** [comment]
I strongly agree that each object should have one required field, if only to state that it exists. 
For the velocity, if any value exists, it is the forward velocity (vx). 

As you mention, for the load, I see a few different common cases we have:

1. Pick & Place: I have walls of identical boxes, that load on the same place on a robot, with no load sensors, but I know the **loadId** / qr-code / item-no that is on the box.
2. Vertical Loading: Material is loaded on the robot, I don't know what it is, it fits within a standard volume (pallet, compartment, etc), but I do have load sensors to know how the **weight** (important for speed / acceleration modulation).
3. Forklift Pallets of Different Sizes: we have some regular sized pallets, some 2x sized pallets, and sometimes those 2x size pallets loaded up, with a high vertical volume, or some very long objects on top (ex sticks of the pallet by 1/2 length off each end). In this case, the **boundingBox** and **loadDimensions** would also be critical for node / path / volumetric reservation. 
4. Horizontal Loading, Underneath: different from carrying a load, the load is already on a cart-like structure, but the robot drives it from underneath. The weight is unknown (though could be gathered from kinematics), the loadId is unknown, the dimensions and bounding box are standard to the cart, and therefore usually standard to the robot in the given installation. There's no value that's particularly useful or providable for this load. But you would know if you're loaded.
5. Horizontal Loading, Tugger: in this case, when we have a tugger train at our facility, the tail of carts is of variable length. There is no collective id, the weight could be kinematically computed as inertia (but not useful). Yet again in this case, the **boundingBox** and **loadDimensions** would be relevant. 

As for what I most frequently encounter, leverage, and implement, I commonly use 1, 2, and 4, because for safety and technology reasons the forklifts (3) and tuggers (5) are operated by humans. 
Regardless, there's no common member to each of these cases, but I would want to know if there is a load or not, period. Currently, in case 4, it's just transmitting the **load** object, as an empty object.

Therefore, I agree with Philip's proposal, that there should at least be an explicit **loaded** (boolean) (required) that is true when there is a load, and that a false value or an empty **load** object is the absense of a load. 
(Biased such that non-loading robots (ex. cleaning robots), do not need to ever include the load, and load carrying robots always have the load object, just specifying whether loaded, or not, and with what known info.)
